daft/io/unity_catalog.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import dataclasses
+
+import unitycatalog
+
+from daft.io import IOConfig, S3Config
+
+
+@dataclasses.dataclass(frozen=True)
+class UnityCatalogTable:
+    table_uri: str
+    io_config: IOConfig | None
+
+
+class UnityCatalog:
+    """Client to access the Unity Catalog
+
+    Unity Catalog is an open-sourced data catalog that can be self-hosted, or hosted by Databricks.
+
+    Example of reading a dataframe from a table in Unity Catalog hosted by Databricks:
+
+    >>> cat = UnityCatalog("https://<databricks_workspace_id>.cloud.databricks.com", token="my-token")
+    >>> table = cat.load_table("my_catalog.my_schema.my_table")
+    >>> df = daft.read_delta_lake(table)
+    """
+
+    def __init__(self, endpoint: str, token: str | None = None):
+        self._client = unitycatalog.Unitycatalog(
+            base_url=endpoint.rstrip("/") + "/api/2.1/unity-catalog/",
+            default_headers={"Authorization": f"Bearer {token}"},
+        )
+
+    def list_catalogs(self) -> list[str]:
+        catalog_names = []
+
+        # Make first request
+        response = self._client.catalogs.list()
+        if response.catalogs is not None:
+            catalog_names.extend([c.name for c in response.catalogs])
+
+        # Exhaust pages
+        while response.next_page_token is not None and response.next_page_token != "":
+            response = self._client.catalogs.list(page_token=response.next_page_token)
+            if response.catalogs is not None:
+                catalog_names.extend([c.name for c in response.catalogs])
+
+        return catalog_names
+
+    def list_schemas(self, catalog_name: str) -> list[str]:
+        schema_names = []
+
+        # Make first request
+        response = self._client.schemas.list(catalog_name=catalog_name)
+        if response.schemas is not None:
+            schema_names.extend([s.full_name for s in response.schemas])
+
+        # Exhaust pages
+        while response.next_page_token is not None and response.next_page_token != "":
```

Comment:
we can prob make this a generic util function that creates an iterator from a method / function using this paradigm

Reply:
Nice, done